domain/application/state/application.go

```diff
@@ -257,6 +242,45 @@ func (st *State) CreateApplication(
 	return appUUID, nil
 }
 
+func (st *State) insertApplicationUnits(
+	ctx context.Context, tx *sqlair.TX,
+	appUUID coreapplication.ID, args application.AddApplicationArg, units []application.AddUnitArg,
+) error {
+	insertUnits := make([]application.InsertUnitArg, len(units))
+	for i, unit := range units {
+		insertUnits[i] = application.InsertUnitArg{
+			UnitName:         unit.UnitName,
+			Constraints:      unit.Constraints,
+			Storage:          args.Storage,
+			StoragePoolKind:  args.StoragePoolKind,
+			StorageParentDir: args.StorageParentDir,
+			UnitStatusArg: application.UnitStatusArg{
+				AgentStatus:    unit.UnitStatusArg.AgentStatus,
+				WorkloadStatus: unit.UnitStatusArg.WorkloadStatus,
+			},
+		}
+	}
+
+	modelType, err := st.GetModelType(ctx)
+	if err != nil {
+		return errors.Errorf("getting model type: %w", err)
+	}
+	if modelType == coremodel.IAAS {
+		for _, arg := range insertUnits {
+			if err := st.insertIAASUnit(ctx, tx, appUUID, arg); err != nil {
+				return errors.Errorf("inserting IAAS unit %q: %w", arg.UnitName, err)
+			}
+		}
+	} else {
+		for _, arg := range insertUnits {
+			if err := st.insertCAASUnit(ctx, tx, appUUID, arg); err != nil {
+				return errors.Errorf("inserting CAAS unit %q: %w", arg.UnitName, err)
+			}
+		}
+	}
```

Comment:
I wonder if we should have `CreateIAASApplication` and `CreateCAASApplication` and just swallow the pill.

Reply:
Again, 100% agreed and mentioned in the PR description but I just wanted to start by doing to units work here and the applications work would come next.